File: code/examples/python/07_citations.py
"""
07 — Citations

Every extracted field carries citation evidence: the exact text snippet
the model relied on, a confidence score, and (for PDFs) the page number.

This example shows two ways to access citations:
  1. Inline — citations are embedded in every record from sift.records()
  2. Drill-down — sift.record(record_id).citations() fetches a single record's map

Requirements:
    uv run --project ../../sdk-python python 07_citations.py
    # SIFTER_API_KEY and SIFTER_API_URL env vars set
"""
from sifter import Sifter

s = Sifter()

# ── Upload a small batch of invoices and wait ─────────────────────────────────

sift = s.create_sift(
    name="Citations demo",
    instructions="Extract supplier, invoice_date, total_amount and currency.",
)
sift.upload("../documents/invoices/INV-2025-001.pdf", on_conflict="replace")
sift.wait()

# ── 1. Inline citations ───────────────────────────────────────────────────────
# Every record returned by .records() already includes a `citations` dict.

records = sift.records()
print(f"Extracted {len(records)} record(s)\n")

for rec in records:
    print(f"Record {rec['id']}  (doc: {rec['document_id']})")
    print(f"  Fields: {rec['extracted_data']}\n")

    citations = rec.get("citations") or {}
    if not citations:
        print("  No citations (document may need reindexing)\n")
        continue

    for field, cit in citations.items():
        page_info = f"  page {cit['page']}" if cit.get("page") else ""
        inferred  = "  (inferred)" if cit.get("inferred") else ""
        confidence = f"{cit['confidence']:.0%}" if cit.get("confidence") is not None else "—"
        print(f"  {field}: \"{cit['source_text']}\"{page_info}{inferred}  [{confidence}]")
    print()

# ── 2. Drill-down via RecordHandle ────────────────────────────────────────────
# Useful when you already have a record_id and don't want to re-fetch all records.

if records:
    record_id = records[0]["id"]
    handle = sift.record(record_id)
    cit_response = handle.citations()

    print(f"── Drill-down citations for {record_id} ──")
    for field, cit in (cit_response.get("citations") or {}).items():
        page_info = f"  page {cit['page']}" if cit.get("page") else ""
        print(f"  {field}: \"{cit['source_text']}\"{page_info}")
    print()

# Cleanup
sift.delete()

File: code/examples/ts/07_citations.ts
/**
 * 07 — Citations
 *
 * Every extracted field carries citation evidence: the exact text snippet
 * the model relied on, a confidence score, and (for PDFs) the page number.
 *
 * This example shows how to read inline citations from sift.records() —
 * they are embedded in every SiftRecord without an extra API call.
 *
 * Requirements:
 *   npx tsx 07_citations.ts
 *   # SIFTER_API_KEY and SIFTER_API_URL env vars set
 */
import { Sifter, type SiftRecord, type Citation } from "@sifter-ai/sdk";
// RecordHandle is returned by sift.record(id) — no separate import needed

const s = new Sifter();

// ── Upload a single invoice and wait ─────────────────────────────────────────

const sift = await s.createSift(
  "Citations demo TS",
  "Extract supplier, invoice_date, total_amount and currency.",
);
await sift.upload("../documents/invoices/INV-2025-001.pdf", { onConflict: "replace" });
await sift.wait();

// ── Read inline citations ─────────────────────────────────────────────────────
// Every record from sift.records() already includes a `citations` map.

const records = await sift.records<SiftRecord>();
console.log(`Extracted ${records.length} record(s)\n`);

for (const rec of records) {
  console.log(`Record ${rec.id}  (doc: ${rec.document_id})`);
  console.log(`  Fields: ${JSON.stringify(rec.extracted_data)}\n`);

  const citations = rec.citations ?? {};
  const fields = Object.keys(citations);

  if (fields.length === 0) {
    console.log("  No citations (document may need reindexing)\n");
    continue;
  }

  for (const field of fields) {
    const cit: Citation = citations[field]!;
    const pageInfo   = cit.page != null ? `  page ${cit.page}` : "";
    const inferred   = cit.inferred ? "  (inferred)" : "";
    const confidence = cit.confidence != null ? `${(cit.confidence * 100).toFixed(0)}%` : "—";
    console.log(`  ${field}: "${cit.source_text}"${pageInfo}${inferred}  [${confidence}]`);
  }
  console.log();
}

// ── Drill-down via record().citations() ──────────────────────────────────────
// Useful when you already have a record_id and don't need to re-fetch all records.

if (records.length > 0) {
  const recordId = records[0]!.id;
  const handle = sift.record(recordId);
  const citResponse = await handle.citations();

  console.log(`── Drill-down citations for ${recordId} ──`);
  for (const [field, cit] of Object.entries(citResponse.citations ?? {})) {
    const pageInfo = cit.page != null ? `  page ${cit.page}` : "";
    console.log(`  ${field}: "${cit.source_text}"${pageInfo}`);
  }
  console.log();
}

// Cleanup
await sift.delete();
